### PR TITLE
tech-debt/aws_backup_plan: update acceptance testing with SDKv2 TypeSet checks

### LINKED ISSUE #14569: tests/resource/aws_backup_plan: TestAccAwsBackupPlan_withLifecycle Failures
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

We recently upgraded to version 2 of the Terraform Plugin SDK, which uses real Terraform binaries and command outputs for handling acceptance testing. `TypeSet` attribute state value checking is (correctly) no longer possible via flatmap or in rare cases where the `.0` `TypeSet` index worked. The `aws_backup_plan` resource contains testing which is dependent on the latter:

```
=== CONT  TestAccAwsBackupPlan_withLifecycle
TestAccAwsBackupPlan_withLifecycle: resource_aws_backup_plan_test.go:191: Check 7/8 error: aws_backup_plan.test: Attribute 'rule.0.lifecycle.0.delete_after' not found
--- FAIL: TestAccAwsBackupPlan_withLifecycle (11.63s)
```

This testing should be updated to the set testing functions, e.g. `TestCheckTypeSetElemNestedAttrs`.

### Affected Resource(s)

* aws_backup_plan

### Definition of Done

- [ ] `TestAccAwsBackupPlan_withLifecycle` testing passes

### References

* [Contributing Guide on Running/Writing Acceptance Testing](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/contributing/running-and-writing-acceptance-tests.md)

### PR DESCRIPTION
**Pull request by anGie44**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #14569

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
tech-debt/aws_backup_plan: update acceptance testing with SDKv2 TypeSet checks
```

Output from acceptance testing:

```
--- PASS: TestAccAwsBackupPlan_disappears (25.80s)
--- PASS: TestAccAwsBackupPlan_basic (29.41s)
--- PASS: TestAccAwsBackupPlan_Rule_CopyAction_CrossRegion (33.19s)
--- PASS: TestAccAwsBackupPlan_Rule_CopyAction_Multiple (33.74s)
--- PASS: TestAccAwsBackupPlan_withRules (68.50s)
--- PASS: TestAccAwsBackupPlan_withRecoveryPointTags (69.41s)
--- PASS: TestAccAwsBackupPlan_withTags (70.33s)
--- PASS: TestAccAwsBackupPlan_Rule_CopyAction_SameRegion (97.06s)
--- PASS: TestAccAwsBackupPlan_withLifecycle (118.37s)

--- PASS: TestTestCheckTypeSetElemAttrPair (0.00s)
--- PASS: TestTestMatchTypeSetElemNestedAttrs (0.00s)
--- PASS: TestTestCheckTypeSetElemNestedAttrs (0.00s)
--- PASS: TestTestCheckTypeSetElemNestedAttrs_empty (0.00s)

```

### CODE REVIEW

File: aws/resource_aws_backup_plan_test.go [hunk 1]
```diff
@@ -31,11 +32,13 @@ func TestAccAwsBackupPlan_basic(t *testing.T) {
...
-					testAccCheckAwsBackupPlanRuleAttr(resourceName, &ruleNameMap, rName, "target_vault_name", rName),
-					testAccCheckAwsBackupPlanRuleAttr(resourceName, &ruleNameMap, rName, "schedule", "cron(0 12 * * ? *)"),
-					testAccCheckAwsBackupPlanRuleAttr(resourceName, &ruleNameMap, rName, "lifecycle.#", "0"),
-					testAccCheckAwsBackupPlanRuleAttr(resourceName, &ruleNameMap, rName, "recovery_point_tags.%", "0"),
+					tfawsresource.TestCheckTypeSetElemNestedAttrs(resourceName, "rule.*", map[string]string{
+						"rule_name":             rName,
+						"target_vault_name":     rName,
+						"schedule":              "cron(0 12 * * ? *)",
+						"lifecycle.#":           "0",
+						"recovery_point_tags.%": "0",
```
Comment (anGie44) [inline aws/resource_aws_backup_plan_test.go:40]: originally this attribute was causing the test to fail similar to what happens when `"lifecycle.0.delete_after":       "0",` is checked below as these attributes are not in state for some reason 🤔 the update in `tfawsresource` should account for these

Comment (bflad) [reply aws/resource_aws_backup_plan_test.go:40]: You may want to check with the SDK team, but I have a feeling that we will not want to support the legacy flatmap `.#` and `.%` syntax checks for missing/empty attributes in anything new until they come up with a "better" way to do rich type checking in the acceptance testing functions. I'd expect at some point we could be able to write something like:

```
"recovery_point_tags": nil,
```

Or

```
"recovery_point_tags": something.EmptyOrMissingMap(),
```

And the framework will do the right thing. 👍  For here and now I think we can drop these "missing attribute" specific checks.

File: aws/resource_aws_backup_plan_test.go [hunk 2]
```diff
@@ -31,11 +32,13 @@ func TestAccAwsBackupPlan_basic(t *testing.T) {
 					testAccMatchResourceAttrRegionalARN(resourceName, "arn", "backup", regexp.MustCompile(`backup-plan:.+`)),
 					resource.TestCheckResourceAttr(resourceName, "name", rName),
 					resource.TestCheckResourceAttr(resourceName, "rule.#", "1"),
-					testAccCheckAwsBackupPlanRuleAttr(resourceName, &ruleNameMap, rName, "rule_name", rName),
```
Comment (bflad) [inline aws/resource_aws_backup_plan_test.go:34]: While we're dropping this custom `TestCheckFunc` I think it'd also be worth getting rid of the `ruleNameMap` handling completely in all the test functions and `testAccCheckAwsBackupPlanExists()`. 👍

File: aws/resource_aws_backup_plan_test.go (hunk 1 above)
Comment (anGie44) [reply aws/resource_aws_backup_plan_test.go]: sounds good! now that you mention it, that legacy support for flatmaps/missing attributes does sound familiar especially going back to when these new helper methods were implemented 👍

File: aws/resource_aws_backup_plan_test.go (hunk 2 above)
Comment (anGie44) [reply aws/resource_aws_backup_plan_test.go:34]: agreed! removed in b6b9297

**Pull request merged**